### PR TITLE
Support seed job creation in Configuration as Code, updated

### PR DESCRIPTION
**Pull request by casz**
It is using a released version and the changes that have happened to the configurator in the meantime.
SeedJob configurator now has access to JCasC secrets.

### CODE REVIEW

File: job-dsl-plugin/src/main/groovy/javaposse/jobdsl/plugin/casc/SeedJobConfigurator.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        return config.asSequence().stream()
+            .map(source -> getActualValue(source, context))
+            .map(source -> getScriptFromSource(source, context, configurator))
+            .map(script -> generateFromScript(script, management))
+            .toArray(GeneratedItems[]::new);
+    }
+
+    @Override
+    public GeneratedItems[] check(CNode config, ConfigurationContext context) throws ConfiguratorException {
+        // Any way to dry-run a job-dsl script ?
```
Comment (daspilker) [inline job-dsl-plugin/src/main/groovy/javaposse/jobdsl/plugin/casc/SeedJobConfigurator.java]: No, not yet.

File: job-dsl-plugin/src/main/groovy/javaposse/jobdsl/plugin/casc/ConfigurableScriptSource.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package javaposse.jobdsl.plugin.casc;
+
+import io.jenkins.plugins.casc.Configurable;
+import io.jenkins.plugins.casc.ConfiguratorException;
+import io.jenkins.plugins.casc.model.CNode;
+
+/**
+ * @author <a href="mailto:nicolas.deloof@gmail.com">Nicolas De Loof</a>
```
Comment (daspilker) [inline job-dsl-plugin/src/main/groovy/javaposse/jobdsl/plugin/casc/ConfigurableScriptSource.java]: Please remove all author tags.

**Comment by casz** [comment]
@daspilker Done would be great to get this into job dsl :crossed_fingers: 

as demonstrated by Matrix auth plugin the configurators can co-exist as long as the ordinal is higher than what is than what configuration-as-code-support-plugin ordinal is :sweat_smile: 

https://github.com/jenkinsci/matrix-auth-plugin/pull/48

**Comment by casz** [comment]
We just merged https://github.com/jenkinsci/configuration-as-code-plugin/pull/845
I will see if I can create a similar PR for removing job dsl from jcasc-support plugin

**Comment by timja** [comment]
There's no incremental build here @casz

**Comment by casz** [comment]
I know, will have to deploy a snapshot :laughing:

**Comment by daspilker** [comment]
@casz Can I release this change anytime or do you need to remove Job DSL from jcasc-support plugin first?

**Comment by casz** [comment]
you can release it anytime @daspilker :raised_hands: it was just if you wanted proof that it works since we have some integrations tests.

The PR would be very similar to https://github.com/jenkinsci/configuration-as-code-plugin/pull/845

**Comment by casz** [comment]
@daspilker let me get the integration tests working... :sweat: Seems we might have trouble

**Comment by daspilker** [comment]
@casz I did a manual test. I never used CasC before, I wanted to see how it works. Quite nice actually 👍 . This Job DSL section does not appear in the reference. Is that expected?

**Comment by casz** [comment]
Oh I just forgot to download all the snapshots... I only deployed the plugin not core...

**Comment by casz** [comment]
That is expected, we could add some sort of support for exporting like storing the sources. Shouldn't be too difficult :sweat:

**Comment by daspilker** [comment]
What do you mean by storing the sources? Must that be done on the CasC side? Will that be available soon?

I would also be fine with adding a wiki page with some examples that we can point to. https://github.com/jenkinsci/job-dsl-plugin/tree/master/docs will be pushed to the [wiki](https://github.com/jenkinsci/job-dsl-plugin/wiki) on every Job DSL release. So you could add a short page to this PR. Maybe you have some existing docs from the support plugin?

**Comment by timja** [comment]
Existing docs are here:
https://github.com/jenkinsci/configuration-as-code-plugin/tree/master/demos/jobs

Seems retrieving from a url or file is undocumented there, it only has inline

**Comment by casz** [comment]
The configurator just has to describe the sources so it can be exported.

**Comment by casz** [comment]
@timja, I can't for the life of me figure out how to access the ScriptSource instances and describe them.

Unless you have some ideas let's script the export.

I'll add some docs

**Comment by casz** [comment]
🎉

**Comment by linux-samurai** [comment]
The release yesterday broke all of our jobs that are using some custom code we wrote in a shared-pipeline-library.  I can provide more details about the custom code if needed, but the error message from the jenkins master is here. Jobs were working up until jenkins master ran auto-updates and pulled this new version down (from 1.72 to 1.73). Please advise. 

```
WARNING: Failed to instantiate Key[type=javaposse.jobdsl.plugin.casc.FromFileScriptSource$DescriptorImpl, annotation=[none]]; skipping this component
com.google.inject.ProvisionException: Unable to provision, see the following errors:

1) Error injecting constructor, java.lang.NoClassDefFoundError: javaposse/jobdsl/plugin/casc/FromFileScriptSource
  at javaposse.jobdsl.plugin.casc.FromFileScriptSource$DescriptorImpl.<init>(FromFileScriptSource.java:28)

Caused by: java.lang.NoClassDefFoundError: javaposse/jobdsl/plugin/casc/FromFileScriptSource

com.google.inject.internal.ProviderToInternalFactoryAdapter$1.call(ProviderToInternalFactoryAdapter.java:46)
        at com.google.inject.internal.InjectorImpl.callInContext(InjectorImpl.java:1103)
        at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
        ... 54 more
```

**Comment by casz** [comment]
Not sure whether you're actually using JCasC or not.

if you're using JCasC you should define your shared libraries inside JCasC.
So JobDSL can access your shared library code I guess.

You seem to have left out the context of where this error message appeared @linux-samurai

**Comment by linux-samurai** [comment]
@casz I did not write our shared library code, and still getting up to speed on Groovy coding, so please bear with me. We are not using the 'configuration as code' plugin, but do a lot of the similar functionality within our shared library that the configuration as code plugin looks to be doing.   

An example of one of our Jenkinsfiles is as follows, which loads the shared library:
```
@Library('company-pipeline-lib') _
archetype('GitOrganizationNameRedacted/jenkins-repository-name-redacted') {
config_select = [ applications: params.deploy_application ]
 }
```


We then have an 'archetype.groovy' in our shared library that takes the input and pulls that declarative pipeline code and injects the 'config_select'. We have a lot of custom DSL methods we wrote that allow us to extend the declarative pipeline a lot, for example we have a 'define' custom method that will take just about any input (string, boolean, key/value pair, array, etc).  So this release of job-dsl plugin, and from the error, this specific PR, broke every job we have that is using the custom DSL methods we have in our shared library.  We have downgraded to 1.72 to get unblocked, but I would like to help flesh out what the root cause is (in our code and how should we improve it or improvements or plugin code improvements) and help improve the code in general, because I wouldn't want us to be locked at this version forever. 

If you could let me know if there is something specific I can look for in our shared library code I can circle back.  

here is another example of a custom method that was written by the original author, in case it helps:```

@NonCPS
def getBuild(jobName, buildId) {
  Jenkins.get().getItemByFullName(jobName)?.getBuild(buildId)
}
```

**Comment by linux-samurai** [comment]
@casz Sorry, I want to explicitly answer your questions, in case they were not laid out fully in my response earlier. 
1. We are not using the config-as-code plugin.
2. This only affects systems where we have job-dsl plugin installed. 
3. For now we have downgraded job-dsl plugin to 1.72.
3. The error was being thrown in a part of the code where we are calling our 'gcloud' custom DSL method. 

(job code)
```
 gcloud 'app deploy', app_yaml, version: appVersion, transform: { it.versions?.first() }
```

In the shared library, the gcloud.groovy file is a bit lengthy, but here is what we are importing. I'd need to review the code a bit more, and get approval to share it.  Initial review doesn't have any glaring sensitivity concerns. 

```
import groovy.json.*
import util.*
```

**Comment by timja** [comment]
Sounds like an optional plugin dependency issue?

**Comment by casz** [comment]
Nah, sounds we have duplicate symbols: https://github.com/jenkinsci/job-dsl-plugin/blob/ff202a73671777f519402e84cd7d4ecdbab7b08e/job-dsl-plugin/src/main/groovy/javaposse/jobdsl/plugin/casc/FromFileScriptSource.java#L27

**Comment by daspilker** [comment]
This has been reported as [JENKINS-57218](https://issues.jenkins-ci.org/browse/JENKINS-57218).

Shouldn't symbols have their descriptor type as namespace? Why would the Pipeline DSL pick the `FromFileScriptSource` in this case? Isn't this a Pipeline problem?

**Comment by timja** [comment]
Would the easiest fix be to rename fromFileScriptSource symbol to be something like?:
```
filePath
```

or 
```
fileLocation
``

**Comment by casz** [comment]
or `jobdslFile` to avoid conflicts all together

**Comment by casz** [comment]
I agree @daspilker it is a problem in the pipeline. It has been reported several times that Symbol lacks more control. 
CC @abayer

**Comment by michelzanini** [comment]
This has broke our builds also. It fails in startup but also on pipeline executions.
I do not have JCasC installed.

**Comment by casz** [comment]
It seems the fix was: #1175 and https://github.com/jenkinsci/structs-plugin/pull/45
Sorry for not marking it optional in this PR completely forgot :laughing:
Lucky we caught a null pointer in structs plugin :raised_hands:

**Comment by jetersen** [comment]
FYI https://github.com/jenkinsci/configuration-as-code-plugin/pull/1394 should fix any ordering to the root elements so the job dsl seed job should load last.

**Pull request merged**